File: solutions/elixir/high-school-sweetheart/1/lib/high_school_sweetheart.ex
defmodule HighSchoolSweetheart do
  def first_letter(name) do
    name
    |> String.trim()
    |> String.first()
  end

  def initial(name) do
    name
    |> first_letter()
    |> String.upcase()
    |> Kernel.<>(".")
  end

  def initials(full_name) do
    [forename, surname] = full_name
    |> String.split(" ")
    initial1 = forename
    |> initial()
    initial2 = surname
    |> initial()

    "#{initial1} #{initial2}"
    # terrible but couldn't get Enum.map() to work
  end

  def pair(full_name1, full_name2) do

    initials1 = full_name1
    |> initials
    initials2 = full_name2
    |> initials

    """
         ******       ******
       **      **   **      **
     **         ** **         **
    **            *            **
    **                         **
    **     #{initials1}  +  #{initials2}     **
     **                       **
       **                   **
         **               **
           **           **
             **       **
               **   **
                 ***
                  *
    """
    
  end
end

File: solutions/php/sweethearts/1/HighSchoolSweetheart.php
<?php

class HighSchoolSweetheart
{
    public function firstLetter(string $name): string
    {
        return trim($name)[0];
    }

    public function initial(string $name): string
    {
        return strtoupper(self::firstLetter($name)) . '.';
    }

    public function initials(string $name): string
    {
        // trim whitespace |> split (explode) on whitespace |> map against initial() |> concatenate w/' '
        return implode(' ', array_values(array_map('self::initial', explode(' ', $name))));
    }

    public function pair(string $sweetheart_a, string $sweetheart_b): string
    {

        $initials = [self::initials($sweetheart_a), self::initials($sweetheart_b)];
        
        return <<<HEART
     ******       ******
   **      **   **      **
 **         ** **         **
**            *            **
**                         **
**     $initials[0]  +  $initials[1]     **
 **                       **
   **                   **
     **               **
       **           **
         **       **
           **   **
             ***
              *
HEART;
    }
}
